File: Extra/Source/tests/integrations.test.js
import test from "node:test";
import assert from "node:assert/strict";
import { execFile } from "node:child_process";
import { promisify } from "node:util";
import { fileURLToPath } from "node:url";
import { Sreon } from "../integrations/sreon.mjs";

const helper = fileURLToPath(new URL("./fixtures/api-helper.mjs", import.meta.url));
const client = (options = {}) => new Sreon(process.execPath, { args: [helper], ...options });

test("Node client serializes concurrent requests and preserves Unicode, category, and cursor", async (t) => {
  const engine = client();
  t.after(() => engine.close());
  const cursor = { source: "media", offset: 24 };
  const data = await Promise.all([engine.search("森林", "images", cursor), engine.search("YouTube"), engine.search("forest")]);
  assert.deepEqual(data.map(item => item.results[0].title), ["森林", "YouTube", "forest"]);
  assert.equal(data[0].category, "images");
  assert.deepEqual(data[0].cursor, cursor);
});

test("valid API errors do not disconnect the Node client", async (t) => {
  const engine = client();
  t.after(() => engine.close());
  await assert.rejects(engine.search("error"), (error) => {
    assert.equal(error.message, "Source unavailable");
    assert.equal(error.code, "SEARCH_UNAVAILABLE");
    assert.equal(error.status, 502);
    return true;
  });
  assert.equal((await engine.search("next")).results[0].title, "next");
});

for (const query of ["null", "wrong-id", "wrong-version", "bad-json", "bad-result", "large-response", "exit"]) {
  test(`Node client safely rejects ${query} responses`, async (t) => {
    const engine = client();
    t.after(() => engine.close());
    await assert.rejects(engine.search(query));
    await assert.rejects(engine.search("later"));
    assert.ok(engine.failure);
  });
}

test("Node client times out a stalled helper and rejects queued requests", async (t) => {
  const engine = client({ timeoutMs: 150 });
  t.after(() => engine.close());
  const results = await Promise.allSettled([engine.search("stall"), engine.search("queued")]);
  for (const result of results) {
    assert.equal(result.status, "rejected");
    assert.match(result.reason.message, /timed out/);
  }
});

test("Node close is idempotent and immediately rejects pending work", async () => {
  const engine = client();
  const rejection = assert.rejects(engine.search("stall"), /closed/);
  engine.close();
  engine.close();
  await rejection;
  await assert.rejects(engine.search("later"), /closed/);
});

test("oversized and unserializable requests never corrupt the Node protocol", async (t) => {
  const engine = client();
  t.after(() => engine.close());
  await assert.rejects(engine.search("x".repeat(17000)), /16 KiB/);
  const cursor = {};
  cursor.circular = cursor;
  await assert.rejects(engine.search("test", "web", cursor), /circular/i);
  assert.equal((await engine.search("ok")).results[0].title, "ok");
});

test("Node client bounds the pending queue", async (t) => {
  const engine = client({ timeoutMs: 10000 });
  t.after(() => engine.close());
  const pending = Array.from({ length: 32 }, () => engine.search("stall").catch(error => error));
  await assert.rejects(engine.search("extra"), /pending searches/);
  engine.close();
  await Promise.all(pending);
});

test("Python client protocol and lifecycle regression tests", async () => {
  const execute = promisify(execFile);
  await execute(process.platform === "win32" ? "python" : "python3", ["tests/integrations_test.py"], { env: { ...process.env, SREON_TEST_NODE: process.execPath }, timeout: 30000 });
});

File: Extra/Source/tests/integrations_test.py
from concurrent.futures import ThreadPoolExecutor
import os
from pathlib import Path
import sys
import unittest

sys.path.insert(0, str(Path(__file__).resolve().parent.parent / "integrations"))
from sreon import Sreon

HELPER = Path(__file__).resolve().parent / "fixtures" / "api-helper.mjs"


class IntegrationTests(unittest.TestCase):
    def client(self, **kwargs):
        engine = Sreon(os.environ["SREON_TEST_NODE"], args=[str(HELPER)], **kwargs)
        self.addCleanup(engine.close)
        return engine

    def test_unicode_and_concurrent_calls(self):
        engine = self.client()
        with ThreadPoolExecutor(max_workers=3) as workers:
            results = list(workers.map(engine.search, ["森林", "YouTube", "forest"]))
        self.assertEqual([item["results"][0]["title"] for item in results], ["森林", "YouTube", "forest"])
        result = engine.search("photos", "photos", {"source": "media", "offset": 24})
        self.assertEqual(result["category"], "photos")
        self.assertEqual(result["cursor"]["offset"], 24)

    def test_recoverable_error(self):
        engine = self.client()
        with self.assertRaisesRegex(RuntimeError, "Source unavailable"):
            engine.search("error")
        self.assertEqual(engine.search("next")["results"][0]["title"], "next")

    def test_protocol_errors_stop_the_helper(self):
        for query in ["null", "wrong-id", "wrong-version", "bad-json", "bad-result", "large-response", "exit"]:
            with self.subTest(query=query):
                engine = self.client()
                with self.assertRaises(RuntimeError):
                    engine.search(query)
                with self.assertRaises(RuntimeError):
                    engine.search("later")
                self.assertIsNotNone(engine.process.poll())

    def test_timeout(self):
        engine = self.client(timeout=0.15)
        with self.assertRaisesRegex(TimeoutError, "timed out"):
            engine.search("stall")
        self.assertIsNotNone(engine.process.poll())

    def test_request_limit_is_recoverable(self):
        engine = self.client()
        with self.assertRaisesRegex(ValueError, "16 KiB"):
            engine.search("x" * 17000)
        self.assertEqual(engine.search("ok")["results"][0]["title"], "ok")

    def test_close_is_idempotent(self):
        engine = self.client()
        engine.close()
        engine.close()
        with self.assertRaisesRegex(RuntimeError, "closed"):
            engine.search("later")

    def test_invalid_timeout(self):
        for timeout in [0, -1, float("inf"), float("nan")]:
            with self.assertRaises(ValueError):
                Sreon("unused", timeout=timeout)


if __name__ == "__main__":
    unittest.main()
